pkg/scheduler/framework/plugins/podtopologyspread/scoring.go

```diff
@@ -0,0 +1,255 @@
+			return
+		}
+		// (1) `node` should satisfy incoming pod's NodeSelector/NodeAffinity
+		// (2) All topologyKeys need to be present in `node`
+		if !predicates.PodMatchesNodeSelectorAndAffinityTerms(pod, node) ||
+			!nodeLabelsMatchSpreadConstraints(node.Labels, state.constraints) {
+			return
+		}
+
+		for _, c := range state.constraints {
+			pair := topologyPair{key: c.topologyKey, value: node.Labels[c.topologyKey]}
+			// If current topology pair is not associated with any candidate node,
+			// continue to avoid unnecessary calculation.
+			if state.topologyPairToPodCounts[pair] == nil {
+				continue
+			}
+
+			// <matchSum> indicates how many pods (on current node) match the <constraint>.
+			matchSum := int64(0)
+			for _, existingPod := range nodeInfo.Pods() {
+				if c.selector.Matches(labels.Set(existingPod.Labels)) {
+					matchSum++
+				}
+			}
+			atomic.AddInt64(state.topologyPairToPodCounts[pair], matchSum)
+		}
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodes), processAllNode)
+
+	cycleState.Write(postFilterStateKey, state)
+	return nil
+}
+
+// Score invoked at the Score extension point.
+// The "score" returned in this function is the matching number of pods on the `nodeName`,
+// it is normalized later.
+func (pl *PodTopologySpread) Score(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, err := pl.sharedLister.NodeInfos().Get(nodeName)
+	if err != nil || nodeInfo.Node() == nil {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v, node is nil: %v", nodeName, err, nodeInfo.Node() == nil))
+	}
+
+	node := nodeInfo.Node()
+	s, err := getPostFilterState(cycleState)
+	if err != nil {
+		return 0, framework.NewStatus(framework.Error, err.Error())
+	}
+
+	// Return if the node is not qualified.
+	if _, ok := s.nodeNameSet[node.Name]; !ok {
+		return 0, nil
+	}
+
+	// For each present <pair>, current node gets a credit of <matchSum>.
+	// And we sum up <matchSum> and return it as this node's score.
+	var score int64
+	for _, c := range s.constraints {
+		if tpVal, ok := node.Labels[c.topologyKey]; ok {
+			pair := topologyPair{key: c.topologyKey, value: tpVal}
+			matchSum := *s.topologyPairToPodCounts[pair]
+			score += matchSum
+		}
+	}
+	return score, nil
+}
+
+// NormalizeScore invoked after scoring all nodes.
+func (pl *PodTopologySpread) NormalizeScore(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, scores framework.NodeScoreList) *framework.Status {
+	s, err := getPostFilterState(cycleState)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	if s == nil {
+		return nil
+	}
+
+	// Calculate the summed <total> score and <minScore>.
+	var minScore int64 = math.MaxInt64
+	var total int64
+	for _, score := range scores {
+		// it's mandatory to check if <score.Name> is present in m.nodeNameSet
+		if _, ok := s.nodeNameSet[score.Name]; !ok {
```

Comment:
why? the score is zero anyways, right?

Reply:
It's "invalid" (as explained https://github.com/kubernetes/kubernetes/pull/86684#discussion_r361819269), which should only be finalized to 0 at L228~L231; otherwise, it would be treated as a node which has 0 matching pods.